### ISSUE TITLE
 v25.0 regression: `C2370` compile error from `map_field.h` when compiling with `cl.exe` and `/permissive-` option

### ISSUE THREAD

**Issue by yukawa** [protocolbuffers/protobuf#14602]
**What version of protobuf and what language are you using?**
Version: v25.0. Looks like a regression in 95d57237b3059548a98da12cdf28807ffcacad5b.
Language: C++

**What operating system (Linux, Windows, ...) and version?**
Windows 11 ver 22H2 (OS Build 22621.2428)

**What runtime / compiler are you using (e.g., python version or gcc version)**
* Microsoft Visual Studio Community 2022 (64-bit) Version 17.7.0
  * Microsoft (R) C/C++ Optimizing Compiler Version 19.37.32822 for x64
  * cmake version 3.26.4-msvc4

**What did you do?**
```
"C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Auxiliary\Build\vcvarsall.bat" amd64_x86
git clone -b v25.0 https://github.com/protocolbuffers/protobuf.git --recursive
cd protobuf
cmake . -DCMAKE_CXX_STANDARD=20 -Dprotobuf_ABSL_PROVIDER=module
cmake --build . -j
```

**What did you expect to see**
Build succeeds.

**What did you see instead?**
Build fails with the following error.
```
printer.cc
C:\Users\builder\source\repos\protobuf\src\google/protobuf/map_field.h(682,66): error C2370: 'google::protobuf::internal::MapField<Derived,Key,T,kKeyFieldType_,kValueFieldType_>::kVTable': redefinition; different storage class (compiling source file C:\Users\builder\source\repos\protobuf\src\google\protobuf\struct.pb.cc) [C:\Users\builder\source\repos\protobuf\libprotobuf.vcxproj]
C:\Users\builder\source\repos\protobuf\src\google/protobuf/map_field.h(671,37): message : see declaration of 'google::protobuf::internal::MapField<Derived,Key,T,kKeyFieldType_,kValueFieldType_>::kVTable' (compiling source file C:\Users\builder\source\repos\protobuf\src\google\protobuf\struct.pb.cc) [C:\Users\builder\source\repos\protobuf\libprotobuf.vcxproj]
  strtod.cc
```

It's failing here.
https://github.com/protocolbuffers/protobuf/blob/6b5d8db01fe47478e8d400f550e797e6230d464e/src/google/protobuf/map_field.h#L671-L683

**Anything else we should know about your project / environment**

Seems that this is a known issue in `cl.exe`.
 * [C2370 is falsely emitted when a static const member is defined with the constexpr specifier outside a template class](https://developercommunity.visualstudio.com/t/c2370-is-falsely-emitted-when-a-static-const-membe/898167)

The tricky thing is that [`/permissive-`](https://learn.microsoft.com/en-us/cpp/build/reference/permissive-standards-conformance) is enabled by default when [`/std:c++20`](https://learn.microsoft.com/en-us/cpp/build/reference/std-specify-language-standard-version) is specified. While developers can still work around this by manually setting `/permissive` only when building protobuf, but it'd be nice if we can avoid this error in protobuf in the first place.

**Comment by ericroy** [comment]
Still not fixed in 25.2
The key to reproducing is that you must compile as c++20 (`CMAKE_CXX_STANDARD=20`)

**Comment by versat** [comment]
> Still not fixed in 25.2 The key to reproducing is that you must compile as c++20 (`CMAKE_CXX_STANDARD=20`)

Looks like we have to wait until 26.0 is released. I hope there will not be any issues again :(
